### PR TITLE
Robolectric extension to support android unit tests

### PR DESCRIPTION
**Pull request by bangarharshit**
Fixes - https://github.com/spekframework/spek/issues/65
I still need to add documentation, unit tests, and code styling but sending an early PR for feedback. I have never written kotlin so it can be rusty.

There are multiple limitations since unlike JUnit, Spek doesn't create a new instance of the unit test. Method level configs and multiple SDK support for tests are not available. I will add the limitations in the doc.

Also, there are lifecycle callbacks which can't be used since it expects a java.lang.reflect.Method. Need to think about it carefully and remove those (they won't be invoked).

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
@bangarharshit you might want to double check which email you used to commit changes, commit and PR author point to different GitHub accounts

To fix that do:

```console
$ cd spek
$ git config user.email "correct@email.com"
$ git commit --amend --reset-author
```

**Comment by bangarharshit** [comment]
Thanks.  Fixed the issue.

File: spek-extensions/robolectric/build.gradle
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/build.gradle]: Why do we need it?

File: spek-extensions/robolectric/build.gradle
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/build.gradle]: Please use `'org.robolectric:robolectric:3.4.2'` form of dependency declaration for consistency

File: spek-extensions/robolectric/build.gradle
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/build.gradle]: Why do you need same dependency declared as `provided`, `compile` and `testRuntime`? 

`api` should be enough, see https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `=`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `=`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `=`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `=`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `annotation?.shadows?.forEach { }`?

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: `=`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Hm, why can it be null here?

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt]: Just enforce it right after initialization :)

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt
Comment (artem-zinnatullin) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt]: seems like you actually want `lateinit` here

File: spek-extensions/robolectric/build.gradle
Comment (raniejade) [inline spek-extensions/robolectric/build.gradle]: We could use the built-in `compileOnly` configuration.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (raniejade) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: nit: `RobolectricInstanceFactory: InstanceFactory`

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (raniejade) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: curious what is this for?

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (raniejade) [inline spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: why do we need `@Transient`?

**Review by raniejade — CHANGES_REQUESTED** [review]

**Comment by bangarharshit** [comment]
@raniejade @artem-zinnatullin Actually the code is a highly modified version of [RobolectricTestRunner.java](https://github.com/robolectric/robolectric/blob/master/robolectric/src/main/java/org/robolectric/RobolectricTestRunner.java#L80) and [SandboxTestRunner.java](https://github.com/robolectric/robolectric/blob/master/junit/src/main/java/org/robolectric/internal/SandboxTestRunner.java).

The big differences from the original are:
1. Original has classloaders for every unit test. Different classloaders are based on different android app configuration (such as sdk=21, 23). In our case, we just throw an error.
2. Moved all the relevant code from callbacks such as JUnit @AfterTest to the correct lifecycle callback.
3. All the limitations mentioned above.


I am not sure how will license work in this case? Or should it be kept in a separate repo?

**Comment by bangarharshit** [comment]
@raniejade saw your comment on the robolectric thread. I am not sure how long will the support on robolectric takes (I can try submitting a PR if someone from community tells the direction). Do you recommend keeping it in a separate repo (still need your review comments)?

If the APIs will not change, we can change the internals later.

File: spek-extensions/robolectric/build.gradle
Comment (bangarharshit) [reply spek-extensions/robolectric/build.gradle]: Done.

File: spek-extensions/robolectric/build.gradle
Comment (bangarharshit) [reply spek-extensions/robolectric/build.gradle]: For some reason, I am getting blockingJunit4classrunner in my classpath resulting in noclassfound error. Need to look into more detail.

File: spek-extensions/robolectric/build.gradle
Comment (bangarharshit) [reply spek-extensions/robolectric/build.gradle]: Done.

File: spek-extensions/robolectric/build.gradle
Comment (bangarharshit) [reply spek-extensions/robolectric/build.gradle]: Unable to use api in the project. Currently used implementation. Will look into it.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Done.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: dependecies are cached by cacheddependecy resolve and it is composed there.  Robolectric also declare it as transient.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Robolectric needs to starts up the Poller SunPKCS11-Darwin thread early, outside of any Robolectric classloader

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Done.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Done.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Done.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: Done.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt]: It will not be null.

File: spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt
Comment (bangarharshit) [reply spek-extensions/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt]: fixed.

**Review by bangarharshit — COMMENTED** [review]
Some comments are still not addressed as I still need to find the exact reason

1. API vs implementation - API not working.
2. Removing Junit4 - It is giving noclassdeffound error.

**Comment by bangarharshit** [comment]
@artem-zinnatullin api vs implementation - api needs the java-library plugin and I don't think it is available in the kotlin-jvm plugin. I have added the java-library plugin. Let me know if it will be an issue for publishing artifacts.

Junit4 is compileOnly in https://github.com/robolectric/robolectric/blob/master/robolectric/build.gradle but it is used in https://github.com/robolectric/robolectric/tree/master/junit module. Classloader is trying to load it when the robolectric package is loaded. Will try to find a way to prevent it from loading (not sure if transitive dependency excludes will be useful). Any solution?

**Comment by raniejade** [comment]
I don't think we could do anything with the `junit` dependency.

File: spek-extension/robolectric/build.gradle
Comment (raniejade) [inline spek-extension/robolectric/build.gradle]: do we need this dependency?

File: spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (raniejade) [inline spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt:47]: Why do we need `@Transient`?

File: spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt
Comment (raniejade) [inline spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt:41]: same here - `@Transient`

**Review by raniejade — CHANGES_REQUESTED** [review]

File: spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt
Comment (bangarharshit) [reply spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricInstanceFactory.kt:47]: Robolectric keeps it to fix issues with other frameworks such as mockito and powermock which also does classloader stuff. I don't the exact internals so I kept it here for safety.

**Comment by bangarharshit** [comment]
The external API still needs some polishing and can use some feedback. Though will prefer to keep these things in separate pull requests.
The current test class:
```kotlin
    object SampleInstanceFactoryTest : InstanceFactory by RobolectricInstanceFactory

    @Test
    fun testUsingRobolectric() {
        @CreateWith(SampleInstanceFactoryTest::class)
        @Config(sdk = intArrayOf(21))
        class RobolectricSpek(val sdkEnvironment: SdkEnvironment, val config: Config, val androidManifest: AndroidManifest ): Spek({
            lateinit var activity: TestActivity1
            registerListener(RobolectricLifeCycleListener(sdkEnvironment, config, androidManifest))
```

Can this be abstracted out? I can create a task and we can do it in another PR.

 Also, it will be great to find ways to support multiple API version for a suite. Can multiple instances be created from Instancefactory?

File: spek-extension/robolectric/src/test/kotlin/robolectric/InstanceFactoryTest.kt
Comment (artem-zinnatullin) [inline spek-extension/robolectric/src/test/kotlin/robolectric/InstanceFactoryTest.kt:34]: I guess we could provide `abstract class RobolectriSpek` to remove this from user's code

**Review by artem-zinnatullin — APPROVED** [review]
LGTM, we can prettify code once we merge it, no blockers from me

**Comment by xian** [comment]
Would it make sense for this to live within the Robolectric project itself?

**Comment by bangarharshit** [comment]
Good point. It will make easier to keep the robolectric code consistent but given the end user APIs are still not very clean we may want to iterate and get more feedback. 

@artem-zinnatullin @raniejade thoughts? Also, if it makes sense to clean up the APIs from end user's perspective even before trying that.

**Comment by xian** [comment]
We've got some significant changes to Robolectric coming up soon that'll need to be mirrored here... we'd be very happy to host this in the Robolectric repo and help maintain it if that makes sense.

**Comment by artem-zinnatullin** [comment]
Yes! @bangarharshit @xian ideally, we'd like to host most of this PR in Robolectric and only have tiny `RobolectricSpek` class here that'll basically attach Robolectric to Spek's lifecycle

**Comment by bangarharshit** [comment]
I can try working on a PR in robolectric this weekend. 

@xian there is common code with Runners such as jar resolver and config generation. How do suggest going around it? 

It will be great if you can check this PR. Do you think we should start abstracting out the code now (before I send the PR) or do it gradually?

**Comment by xian** [comment]
@bangarharshit I'd say keep the duplicated code for now; when you send us a PR to Robolectric I'll refactor to use common code. Thanks!

**Comment by bangarharshit** [comment]
Sorry super busy with a deadline this weekend. I will be sending the PR to Robolectric surely be next weekend.

File: spek-extension/robolectric/build.gradle
Comment (jaredsburrows) [inline spek-extension/robolectric/build.gradle:8]: Since you are using `java-library`, can you change the `compile` to `implementation`?

File: spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt
Comment (jaredsburrows) [inline spek-extension/robolectric/src/main/kotlin/org/spekframework/spek2/robolectric/RobolectricLifeCycleListener.kt:111]: Can you make this a `private const val` in a `companion object`?

**Comment by jaredsburrows** [comment]
Great work @bangarharshit!

**Review by jaredsburrows — APPROVED** [review]

**Comment by bangarharshit** [comment]
I started porting it to the robolectric repo. Are there any artifacts which point to junit-platform-test-engine-support in 1.x branch. I don't see it on maven central - https://mvnrepository.com/artifact/org.jetbrains.spek. 

I am currently using the same in my code and it can be pretty useful since I don't have to set up the boilerplate in AbstractJUnitTestEngineTest

**Comment by raniejade** [comment]
@bangarharshit do you mean `2.x`?

**Comment by bangarharshit** [comment]
Are there any artifacts of 2.0? Couldn't find anything on maven central

**Comment by raniejade** [comment]
No, not yet. We have to deal with #317 before we can publish anything. @artem-zinnatullin do you have an ETA for it?

**Comment by bangarharshit** [comment]
Cool spek 2 has the correct artifact available, runtime-common. I will probably wait for 2.x but if it is taking longer then maybe think of nightly releases. Thanks

**Comment by bangarharshit** [comment]
@raniejade any luck with 2.x release given the blocker is resolved.

**Comment by raniejade** [comment]
#317 is resolved now, but we haven't discuss about the dev builds yet. So sorry for the delay, can you download the repo and just run `distribution:publishToMavenLocal` for now? That should allow you to develop locally.

**Comment by bangarharshit** [comment]
Any update. Are you ok about keeping it in the spek repo for now as the final user interface still needs some work? This way it keeps moving with 2.x and once the interfaces are finalized while keeping the user interfaces here we can move the abstracted code to robolectric.

**Comment by raniejade** [comment]
@bangarharshit We are working on publishing the snapshots, please follow #358.

**Comment by bangarharshit** [comment]
Abandoning this since spek-2.0 is out. Will open a PR in robolectric itself.

**Comment by joelhandwell** [comment]
@bangarharshit Is PR opened in robolectric ? If so, can you point us to the url? (I was not able to find.)

**Comment by bangarharshit** [comment]
Not yet. I will be working on it on the next weekend.

**Comment by Townk** [comment]
@bangarharshit sorry to bump this, but do you have the PR link to post here?

**Comment by sweetleon** [comment]
:1 I'd love to see this get into Robolectric.

**Pull request closed without merging**
